docs/usage/design.rst

```diff
@@ -4,23 +4,121 @@ Design of SCAutolib
+of the library models, except for model's **properties**.
+In other words, models are not aware of other models, but only of their properties.
+This is made to clarify the implementation approach of the library: everything that model requires for it's functionality is provided by the Controller.
+For example, if you want to create a new LocalCA, you need to create a new CNF file for this CA before calling the ``setup`` method on LocalCA object
+
+.. code-block:: python
+
+    ca_root_dir = "/etc/SCAutolib/ca"
+
+    cnf_file = SCAutolib.models.file.OpensslCnf("/tmp/ca.cnf", "CA", ca_root_dir)
+    cnf_file.create()
+    cnf_file.save()
+
+    local_ca = SCAutolib.models.ca.CA.LocalCA(root_dir=ca_root_dir, cnf=cnf_file)
+    local_ca.setup()
+
+This implementation would help to avoid situation with vague call of some method
+on unexpected place (model).
+
+Some models have overwritten ``__dict__`` property so they can be serialized to JSON format and stored in the file.
+Overwriting ``__dict__`` is required because not every attribute of the model is JSON serializable (e.g. pathlib.Path object).
+After serialization, the model can be restored (loaded) via corresponding ``load`` method.
+Example:
+
+.. code-block:: python
+
+    class BaseUser:
+        @staticmethod
+        def load(json_file, **kwargs):
+            with json_file.open("r") as f:
+                cnt = json.load(f)
+            if cnt["local"]:
+                user = User(local=cnt["local"],
+                            username=cnt["username"],
+                            password=cnt["password"], ...)
+            else:
+               # IPAUser is loaded similarly to LocalUser
+               ...
+            return user
+
+    class User(BaseUser):
+        ...
+        @property
+        def __dict__(self):
+            dict_ = super().__dict__.copy()
+            for k, v in dict_.items():
+                if type(v) in (PosixPath, Path):
+                    dict_[k] = str(v)
+
+            if self._card:
+                dict_["_card"] = str(self._card.dump_file)
+            return dict_
+
+Example of usage:
+
+.. code-block:: python
+
+    user = User(...)
+    with user.dump_file("w") as f:
+        json.dump(user.__dict__, f)
+    loaded_user = BaseUser.load(user.dump_file)
+
+``dump_file`` is an attribute of the model that defines path to the file where the model is serialized.
+
+Controller
+###########
+
+Controller is a kind of glue between models all models and the view.
+While models are providing low-level atomic functionality, the controller is responsible for the high-level logic.
+Controller calls method in correct order along with creating all necessary files and objects.
+
+Method of the Controller are designed to be called from the View or from other code that would do setup of the system.
+But it is not aimed to be called from the tests just because it is an overkill for them.
```

Comment:
Is "just because it is an overkill" the primary problem?

Reply:
Actually, in simple words, yes